internal/pkg/agent/application/actions/handlers/handler_action_upgrade.go

```diff
@@ -30,13 +33,38 @@ func NewUpgrade(log *logger.Logger, coord *coordinator.Coordinator) *Upgrade {
 	}
 }
 
-// Handle handles UPGRADE action.
-func (h *Upgrade) Handle(ctx context.Context, a fleetapi.Action, _ acker.Acker) error {
+// Handle handles UPGRADE action.  Returns immediately and the actual
+// upgrade happens asynchronously.  This allows for downloads to
+// happen without blocking updates.  If a second UPGRADE action is
+// received before the first is completed, then an error is returned.
+func (h *Upgrade) Handle(ctx context.Context, a fleetapi.Action, ack acker.Acker) error {
 	h.log.Debugf("handlerUpgrade: action '%+v' received", a)
 	action, ok := a.(*fleetapi.ActionUpgrade)
 	if !ok {
 		return fmt.Errorf("invalid type, expected ActionUpgrade and received %T", a)
 	}
-
-	return h.coord.Upgrade(ctx, action.Version, action.SourceURI, action)
+	go func() {
+		h.m.Lock()
+		if h.bkgAction != nil {
+			h.log.Infof("upgrade to version %s already running in background", h.bkgAction.Version)
+			h.m.Unlock()
+			return
+		}
+		h.bkgAction = action
+		h.m.Unlock()
+		h.log.Infof("starting upgrade to version %s in background", action.Version)
+		if err := h.coord.Upgrade(ctx, action.Version, action.SourceURI, action); err != nil {
```

Comment:
Would it make more sense to move the `bkgAction` into the `coord.Upgrade` to ensure that two upgrade actions cannot happen at the same time from 2 different paths?

With this logic at the moment it could be possible to start an upgrade action from Fleet as well as locally over the control protocol. For safety reasons I think it would be better to have the coordinator manage this coordination instead of placing that logic in the dispatch action.

Reply:
good idea.  I moved it and the code is simpler.

Reply:
Thanks for moving it.